common/src/com/thoughtworks/go/buildsession/TaskCommandExecutor.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2017 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.buildsession;
+
+import com.thoughtworks.go.domain.BuildCommand;
+import com.thoughtworks.go.domain.JobResult;
+
+import java.time.Duration;
+import java.time.Instant;
+
+import static com.thoughtworks.go.util.command.ConsoleLogTags.*;
+import static java.lang.String.format;
+
+/**
+ * Executes a task, reporting its description and result
+ */
+public class TaskCommandExecutor implements BuildCommandExecutor {
```

Comment:
After having seen the rest of the PR and parts such as [this](https://github.com/gocd/gocd/pull/3520/files#diff-f21aae1907b779d5291d580fc02aa492R103), I see what you're trying to do. But, the whole point of the BuildSession work was to come up with composable units, using which we can build a full "job" or a "build plan". By moving to specialized executors, it seems less pure - and other agent implementations will need to keep changing this.

Are you sure about this? Do you really think that these two are real abstractions in this grammar? I mean, "task" and "jobResult"? Or, does it just make it easier to write the build plan?